File: backend/app/presentation/schemas/research_note.py
from datetime import date, datetime

from pydantic import BaseModel, ConfigDict, Field


class ResearchNoteCreateRequest(BaseModel):
    project_id: str
    title: str = Field(min_length=1, max_length=255)
    content: str | None = None
    status: str = "draft"
    owner_member_id: int
    written_date: date | None = None
    reviewer_member_id: int | None = None
    reviewed_date: date | None = None
    last_updated_by: int | None = None


class ResearchNoteUpdateRequest(BaseModel):
    title: str | None = Field(default=None, min_length=1, max_length=255)
    content: str | None = None
    status: str | None = None
    owner_member_id: int | None = None
    written_date: date | None = None
    reviewer_member_id: int | None = None
    reviewed_date: date | None = None
    last_updated_by: int | None = None


class ResearchNoteApprovalRequest(BaseModel):
    comment: str | None = None


class ResearchNoteAssignmentRequest(BaseModel):
    author_member_id: int
    reviewer_member_id: int | None = None


class ResearchNoteResponse(BaseModel):
    id: str
    created_at: datetime
    updated_at: datetime
    project_id: str
    title: str
    content: str | None
    status: str
    owner_member_id: int
    written_date: date | None
    reviewer_member_id: int | None
    reviewed_date: date | None
    last_updated_by: int | None
    is_deleted: bool


class ResearchNoteApprovalEventResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    created_at: datetime
    note_id: str
    document_id: str | None
    document_revision_id: int | None
    actor_user_id: int
    actor_member_id: int | None
    event_type: str
    comment: str | None
    signature_snapshot_id: int | None

File: frontend/src/api/researchNote.ts
import { apiFetch } from "./client";

export type ResearchNote = {
  id: string;
  created_at: string;
  updated_at: string;
  project_id: string;
  title: string;
  content: string | null;
  status: string;
  owner_member_id: number;
  written_date: string | null;
  reviewer_member_id: number | null;
  reviewed_date: string | null;
  last_updated_by: number | null;
  is_deleted: boolean;
};

export type ResearchNoteApprovalEvent = {
  id: number;
  created_at: string;
  note_id: string;
  document_id: string | null;
  document_revision_id: number | null;
  actor_user_id: number;
  actor_member_id: number | null;
  event_type: string;
  comment: string | null;
  signature_snapshot_id: number | null;
};

export type CreateResearchNotePayload = {
  project_id: string;
  title: string;
  content?: string;
  owner_member_id: number;
  written_date?: string;
  reviewer_member_id?: number;
  reviewed_date?: string;
  last_updated_by?: number;
};

async function parseResearchNoteError(response: Response, fallback: string): Promise<string> {
  try {
    const payload = await response.json();
    if (typeof payload?.detail === "string") return payload.detail;
  } catch {
    return fallback;
  }
  return fallback;
}

export async function listResearchNotes(projectId: string): Promise<ResearchNote[]> {
  const response = await apiFetch(`/research-notes?project_id=${projectId}`);
  if (!response.ok) throw new Error(await parseResearchNoteError(response, "Failed to fetch research notes"));
  return response.json();
}

export async function createResearchNote(
  payload: CreateResearchNotePayload
): Promise<ResearchNote> {
  const response = await apiFetch("/research-notes", {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });
  if (!response.ok) throw new Error(await parseResearchNoteError(response, "Failed to create research note"));
  return response.json();
}

export async function getResearchNote(noteId: string): Promise<ResearchNote> {
  const response = await apiFetch(`/research-notes/${noteId}`);
  if (!response.ok) throw new Error(await parseResearchNoteError(response, "Failed to fetch research note detail"));
  return response.json();
}

export async function updateResearchNote(
  noteId: string,
  payload: {
    title?: string;
    content?: string;
    owner_member_id?: number;
    written_date?: string;
    reviewer_member_id?: number;
    reviewed_date?: string;
    last_updated_by?: number;
  }
): Promise<ResearchNote> {
  const response = await apiFetch(`/research-notes/${noteId}`, {
    method: "PUT",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });
  if (!response.ok) throw new Error(await parseResearchNoteError(response, "Failed to update research note"));
  return response.json();
}

export async function assignResearchNoteMembers(
  noteId: string,
  payload: {
    author_member_id: number;
    reviewer_member_id?: number | null;
  }
): Promise<ResearchNote> {
  const response = await apiFetch(`/research-notes/${noteId}/assignments`, {
    method: "PUT",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify(payload),
  });
  if (!response.ok) throw new Error(await parseResearchNoteError(response, "Failed to assign research note members"));
  return response.json();
}

export async function deleteResearchNote(noteId: string): Promise<void> {
  const response = await apiFetch(`/research-notes/${noteId}`, {
    method: "DELETE",
  });
  if (!response.ok) throw new Error(await parseResearchNoteError(response, "Failed to delete research note"));
}

async function postApprovalAction(noteId: string, action: string, comment?: string): Promise<ResearchNoteApprovalEvent> {
  const response = await apiFetch(`/research-notes/${noteId}/${action}`, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ comment: comment ?? null }),
  });
  if (!response.ok) throw new Error(await parseResearchNoteError(response, `Failed to ${action} research note`));
  return response.json();
}

export function submitResearchNote(noteId: string, comment?: string): Promise<ResearchNoteApprovalEvent> {
  return postApprovalAction(noteId, "submit", comment);
}

export function approveResearchNote(noteId: string, comment?: string): Promise<ResearchNoteApprovalEvent> {
  return postApprovalAction(noteId, "approve", comment);
}

export function rejectResearchNote(noteId: string, comment: string): Promise<ResearchNoteApprovalEvent> {
  return postApprovalAction(noteId, "reject", comment);
}

export function reopenResearchNote(noteId: string, comment: string): Promise<ResearchNoteApprovalEvent> {
  return postApprovalAction(noteId, "reopen", comment);
}

export async function listResearchNoteApprovalEvents(noteId: string): Promise<ResearchNoteApprovalEvent[]> {
  const response = await apiFetch(`/research-notes/${noteId}/approval-events`);
  if (!response.ok) throw new Error(await parseResearchNoteError(response, "Failed to fetch approval events"));
  return response.json();
}
